File: app/server/notify/channels/pushplus.ts
/**
 * PushPlus 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'pushplus' });

export const pushplusChannel: NotifyChannel = {
    name: 'pushplus',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('PUSH_PLUS_TOKEN')) {
                resolve({ success: false, message: 'PUSH_PLUS_TOKEN 未配置' });
                return;
            }

            const PUSH_PLUS_TOKEN = getConfig('PUSH_PLUS_TOKEN');
            const PUSH_PLUS_USER = getConfig('PUSH_PLUS_USER');
            const PUSH_PLUS_TEMPLATE = getConfig('PUSH_PLUS_TEMPLATE') || 'html';
            const PUSH_PLUS_CHANNEL = getConfig('PUSH_PLUS_CHANNEL') || 'wechat';
            const PUSH_PLUS_WEBHOOK = getConfig('PUSH_PLUS_WEBHOOK');
            const PUSH_PLUS_CALLBACKURL = getConfig('PUSH_PLUS_CALLBACKURL');
            const PUSH_PLUS_TO = getConfig('PUSH_PLUS_TO');

            // 默认为html, 不支持plaintext
            const formattedDesp = desp.replace(/[\n\r]/g, '<br>');

            const url = 'https://www.pushplus.plus/send';
            const body = {
                token: PUSH_PLUS_TOKEN,
                title: text,
                content: formattedDesp,
                topic: PUSH_PLUS_USER,
                template: PUSH_PLUS_TEMPLATE,
                channel: PUSH_PLUS_CHANNEL,
                webhook: PUSH_PLUS_WEBHOOK,
                callbackUrl: PUSH_PLUS_CALLBACKURL,
                to: PUSH_PLUS_TO,
            };

            $.post({ url, json: body }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'PushPlus 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    const result = data as { code?: number; msg?: string; data?: string };
                    if (result?.code === 200) {
                        logger.info({ flowId: result.data }, 'PushPlus 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else {
                        logger.warn({ body: data }, 'PushPlus 发送失败');
                        resolve({ success: false, message: result?.msg || '发送失败' });
                    }
                }
            });
        });
    }
};

if (hasConfig('PUSH_PLUS_TOKEN')) {
    pushplusChannel.enabled = true;
}

export default pushplusChannel;

File: app/server/internal/notify/channels/pushplus.go
package channels

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type PushPlus struct {
	enabled bool
}

func (c *PushPlus) Name() string            { return "pushplus" }
func (c *PushPlus) Enabled() bool           { return c.enabled }
func (c *PushPlus) SetEnabled(enabled bool) { c.enabled = enabled }

func (c *PushPlus) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("PUSH_PLUS_TOKEN") {
		return notify.NotifyResult{Success: false, Message: "PUSH_PLUS_TOKEN 未配置"}
	}

	token := notify.GetConfig("PUSH_PLUS_TOKEN")
	user := notify.GetConfig("PUSH_PLUS_USER")
	template := notify.GetConfig("PUSH_PLUS_TEMPLATE")
	if template == "" {
		template = "html"
	}
	channel := notify.GetConfig("PUSH_PLUS_CHANNEL")
	if channel == "" {
		channel = "wechat"
	}
	webhook := notify.GetConfig("PUSH_PLUS_WEBHOOK")
	callbackURL := notify.GetConfig("PUSH_PLUS_CALLBACKURL")
	to := notify.GetConfig("PUSH_PLUS_TO")

	// Replace newlines with <br> for HTML template
	formattedDesp := strings.NewReplacer("\r\n", "<br>", "\r", "<br>", "\n", "<br>").Replace(desp)

	url := "https://www.pushplus.plus/send"
	body := map[string]string{
		"token":       token,
		"title":       text,
		"content":     formattedDesp,
		"topic":       user,
		"template":    template,
		"channel":     channel,
		"webhook":     webhook,
		"callbackUrl": callbackURL,
		"to":          to,
	}

	resp, err := notify.HTTPPost(url, notify.HttpRequestOptions{JSON: body})
	if err != nil {
		slog.Error("PushPlus 发送失败", "err", err)
		return notify.NotifyResult{Success: false, Message: fmt.Sprintf("发送失败: %v", err), Err: err}
	}

	var result struct {
		Code float64 `json:"code"`
		Msg  string  `json:"msg"`
		Data string  `json:"data"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &result); err != nil {
		slog.Error("PushPlus 解析响应失败", "err", err)
		return notify.NotifyResult{Success: false, Message: "解析响应失败", Err: err}
	}

	if result.Code == 200 {
		slog.Info("PushPlus 发送成功", "flowId", result.Data)
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	msg := result.Msg
	if msg == "" {
		msg = "发送失败"
	}
	slog.Warn("PushPlus 发送失败", "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: msg}
}

func init() {
	ch := &PushPlus{}
	if notify.HasConfig("PUSH_PLUS_TOKEN") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
